File: kmp-algorithm/KmpAlgorithm.kt
fun computeLPS(pattern: String): IntArray {
    val m = pattern.length
    val lps = IntArray(m)
    var length = 0
    var i = 1
    
    while (i < m) {
        if (pattern[i] == pattern[length]) {
            length++
            lps[i] = length
            i++
        } else {
            if (length != 0) {
                length = lps[length - 1]
            } else {
                lps[i] = 0
                i++
            }
        }
    }
    return lps
}

fun kmpSearch(text: String, pattern: String): List<Int> {
    val n = text.length
    val m = pattern.length
    
    if (m == 0) return emptyList()
    
    val lps = computeLPS(pattern)
    val matches = mutableListOf<Int>()
    
    var i = 0
    var j = 0
    while (i < n) {
        if (pattern[j] == text[i]) {
            i++
            j++
        }
        
        if (j == m) {
            matches.add(i - j)
            j = lps[j - 1]
        } else if (i < n && pattern[j] != text[i]) {
            if (j != 0) {
                j = lps[j - 1]
            } else {
                i++
            }
        }
    }
    
    return matches
}

fun main() {
    val text = "ABABDABACDABABCABCABCABCABC"
    val pattern = "ABABCABCABCABC"
    val matches = kmpSearch(text, pattern)
    println("Pattern found at indices: $matches")
}

File: kmp-algorithm/kmp_algorithm.rb
def compute_lps(pattern)
  lps = Array.new(pattern.length, 0)
  len = 0
  i = 1
  
  while i < pattern.length
    if pattern[i] == pattern[len]
      len += 1
      lps[i] = len
      i += 1
    else
      if len != 0
        len = lps[len - 1]
      else
        lps[i] = 0
        i += 1
      end
    end
  end
  lps
end

def kmp_search(text, pattern)
  lps = compute_lps(pattern)
  i = j = 0
  matches = []
  
  while i < text.length
    if pattern[j] == text[i]
      i += 1
      j += 1
    end
    
    if j == pattern.length
      matches << i - j
      j = lps[j - 1]
    elsif i < text.length && pattern[j] != text[i]
      j != 0 ? j = lps[j - 1] : i += 1
    end
  end
  matches
end

puts kmp_search("ABABDABACDABABCABCABCABCABC", "ABABCABCABCABC").inspect